src/worker_utils.ts

```diff
@@ -0,0 +1,62 @@
+import { Pool, PoolClient } from "pg";
+import { WorkerUtilsOptions, TaskOptions, Job } from "./interfaces";
+import { makeWithPgClientFromPool, makeAddJob } from "./helpers";
+import { migrate } from "./migrate";
+import { defaultLogger } from "./logger";
+import { withReleasers, assertPool } from "./runner";
+
+const processPublisherOptions = async (options: WorkerUtilsOptions) => {
+  const { logger = defaultLogger } = options;
+  return withReleasers(async (releasers, _) => {
+    const pgPool: Pool = await assertPool(options, releasers, logger);
+
+    const withPgClient = makeWithPgClientFromPool(pgPool);
+
+    // Migrate
+    await withPgClient(client => migrate(client));
+
+    return { pgPool, withPgClient };
+  });
+};
+
+export class WorkerUtils {
+  private dbTools: Promise<{
+    withPgClient: <T>(
+      callback: (pgClient: PoolClient) => Promise<T>
+    ) => Promise<T>;
+    pgPool: Pool;
+  }>;
+
+  constructor(options: WorkerUtilsOptions) {
+    this.dbTools = processPublisherOptions(options);
+  }
+
+  public async addJob(
+    identifier: string,
+    payload: any = {},
+    options: TaskOptions = {}
+  ): Promise<Job> {
+    const { withPgClient } = await this.dbTools;
+    const addJob = makeAddJob(withPgClient);
```

Comment:
Let's go with a different implementation; the `addJob` function should enqueue promises (maybe via the `deferred` pattern?) and the constructor should trigger a "startup routine" which awaits for the dbTools, then overwrites the `addJob`/etc methods with the new methods, and then replays all the deferred actions. That way once startup is complete all calls to `addJob` will be directly to the underlying `addJob` function with no `await this.dbTools` overhead.

Though TBH this is a performance optimisation, the API is fine and that's what really matters initially.

Reply:
I got rid of the class entirely in the end; turns out that WorkerUtils is really just a helpers object.